### PR TITLE
Add slice functions

### PR DESCRIPTION
**Pull request by noriakis**
Hello, I added the following functions to `dplyr_methods.R`. I would be grateful if you could review the changes.

- `slice_head()`
- `slice_tail()`
- `slice_sample()`
- `slice_min() `
- `slice_max()`

### CODE REVIEW

File: R/dplyr_methods.R [hunk 1]
```diff
@@ -507,6 +507,208 @@ slice.SingleCellExperiment <- function(.data, ..., .by=NULL, .preserve=FALSE) {
...
+        select(-everything(), {{ by }}) |>
+        rowid_to_column(var='row_number___')  |>
+        slice_head(..., n=n, prop=prop, by={{ by }}) |>
+        pull(row_number___)
+
+    if (length(idx) == 0) {
+        stop("tidySingleCellExperiment says:",
+            " the resulting data container is empty.",
+            " Seurat does not allow for empty containers.")
+    }
```
Comment (stemangiola) [inline R/dplyr_methods.R:587]: Look for "Seurat" in all files, and please replace ;)

**Review by stemangiola — CHANGES_REQUESTED** [review]

File: R/dplyr_methods.R (hunk 1 above)
Comment (stemangiola) [reply R/dplyr_methods.R:587]: I believe an empty SCE is completely fine. Not sure this check is needed.

Comment (noriakis) [reply R/dplyr_methods.R]: Thank you very much for your prompt review. I apologize that the part of the message remained the same as `tidyseurat`. Also, to accept the empty SCE, all the checks for zero length ID in the slice functions are removed.

**Review by stemangiola — CHANGES_REQUESTED** [review]
Thanks! Can I please ask to add the 0-length SCE to the uni tests? As we start with our standard SCE and we slice to result to 0 rows, and see what happens (within the unit tests)

**Comment by noriakis** [comment]
Thank you for the guidance. I have added the tests to expect `ncol()=0`. All the functions run without error or warnings, except for `slice_sample` producing the following warning.

```
> pbmc_small |> slice_sample(n=0)
# A SingleCellExperiment-tibble abstraction: 0 × 17
# Features=230 | Cells=0 | Assays=counts, logcounts
# ℹ 17 variables: .cell <chr>, orig.ident <fct>, nCount_RNA <dbl>, nFeature_RNA <int>, RNA_snn_res.0.8 <fct>,
#   letter.idents <fct>, groups <chr>, RNA_snn_res.1 <fct>, file <chr>, ident <fct>, PC_1 <dbl>, PC_2 <dbl>,
#   PC_3 <dbl>, PC_4 <dbl>, PC_5 <dbl>, tSNE_1 <dbl>, tSNE_2 <dbl>
Warning message:
In max(count_cells$n) : no non-missing arguments to max; returning -Inf
```

**Comment by stemangiola** [comment]
> Thank you for the guidance. I have added the tests to expect `ncol()=0`. All the functions run without error or warnings, except for `slice_sample` producing the following warning.
> 
> ```
> > pbmc_small |> slice_sample(n=0)
> # A SingleCellExperiment-tibble abstraction: 0 × 17
> # Features=230 | Cells=0 | Assays=counts, logcounts
> # ℹ 17 variables: .cell <chr>, orig.ident <fct>, nCount_RNA <dbl>, nFeature_RNA <int>, RNA_snn_res.0.8 <fct>,
> #   letter.idents <fct>, groups <chr>, RNA_snn_res.1 <fct>, file <chr>, ident <fct>, PC_1 <dbl>, PC_2 <dbl>,
> #   PC_3 <dbl>, PC_4 <dbl>, PC_5 <dbl>, tSNE_1 <dbl>, tSNE_2 <dbl>
> Warning message:
> In max(count_cells$n) : no non-missing arguments to max; returning -Inf
> ```

Great. If this warning comes from `tidySingleCellExperiment` we should take care of it.

**Comment by noriakis** [comment]
Thank you for the review. I have committed the changes to overcome the warning in `slice_sample`.
I would be grateful if you could review the relevant part.

File: R/dplyr_methods.R [hunk 2]
```diff
@@ -546,8 +546,10 @@ slice_sample.SingleCellExperiment <- function(.data, ..., n=NULL,
         select(!!c_(.data)$symbol) %>%
         count(!!c_(.data)$symbol)
 
+    .max_cell_count <- ifelse(nrow(count_cells)==0, 0, max(count_cells$n))
+
     # If repeated cells due to replacement
-    if (count_cells$n |> max() |> gt(1)){
+    if (.max_cell_count |> gt(1)){
```
Comment (stemangiola) [inline R/dplyr_methods.R:552]: looks good, does the warning appear in `tidyseurat` as well?

Comment (noriakis) [reply R/dplyr_methods.R:552]: Thank you for checking. Seems to be yes, although it results in an error. Should I also replace the part for `tidyseurat`?

```
> pbmc_small |> slice_sample(n=0)
Error: No cells found
In addition: Warning message:
In max(count_cells$n) : no non-missing arguments to max; returning -Inf
```

**Review by stemangiola — APPROVED** [review]

File: R/dplyr_methods.R (hunk 2 above)
Comment (stemangiola) [reply R/dplyr_methods.R:552]: @noriakis  Yes, please fix the error in `tidyseurat` and write a unit test for it.

If this is relative to an empty object I thought you had a `stop()` call already.

Comment (noriakis) [reply R/dplyr_methods.R:552]: Thank you for your reply. The error is related to an empty object coming from `stop()` call in `subset.Seurat()` as you pointed out, so I will just fix the warnings.

**Pull request merged**